wgpu-hal/src/dx12/suballocation.rs

```diff
@@ -0,0 +1,307 @@
+use native::WeakPtr;
+use parking_lot::Mutex;
+use std::ptr;
+use winapi::{
+    um::{
+        d3d12::{self, ID3D12Resource},
+        winnt::HRESULT,
+    },
+    Interface,
+};
+
+// This exists to work around https://github.com/gfx-rs/wgpu/issues/3207
+// Currently this will work the older, slower way if the windows_rs feature is disabled,
+// and will suballocate buffers using gpu_allocator if the windows_rs feature is enabled.
+
+#[cfg(feature = "windows_rs")]
+use gpu_allocator::{
+    d3d12::{AllocationCreateDesc, ToWinapi, ToWindows},
+    MemoryLocation,
+};
+
+// TODO: find the exact value
+// TODO: figure out if this is even needed?
+#[allow(unused)] // TODO: Exists until windows-rs is standard, then it can probably be removed?
+pub(crate) const D3D12_HEAP_FLAG_CREATE_NOT_ZEROED: u32 = d3d12::D3D12_HEAP_FLAG_NONE;
+
+#[derive(Debug)]
+pub(crate) struct GpuAllocatorWrapper {
+    #[cfg(feature = "windows_rs")]
+    pub(crate) allocator: gpu_allocator::d3d12::Allocator,
+    #[cfg(not(feature = "windows_rs"))]
+    #[allow(unused)]
+    pub(crate) allocator: (),
+}
```

Comment:
Lets split this into two large blocks, one for with-windows-rs and one for without, either large cfg blocks or two inline modules, then just expose the same interface from both modules. As it stands you have a bunch of unrelated code mashed together which makes it very hard to follow for someone who doesn't know what's going on.

Reply:
I went with #[cfg] blocked inline modules, seemed cleaner to me.

Reply:
I think having them inline is fine for when there's a decent overlap in the code paths, but when they're basically completely different, it helps to separate them